base_layer/wallet/src/util/encryption.rs

```diff
@@ -50,91 +48,118 @@ pub trait Encryptable<C> {
+        // Encrypt a message
         let plaintext = b"The quick brown fox was annoying".to_vec();
         let key = GenericArray::from_slice(b"an example very very secret key.");
-        let cipher = Aes256Gcm::new(key);
+        let cipher = XChaCha20Poly1305::new(key);
 
         let ciphertext = encrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), plaintext.clone()).unwrap();
+
+        // Check the ciphertext size, which we rely on for later tests
+        // It should extend the plaintext size by the nonce and tag sizes
+        assert_eq!(
+            ciphertext.len(),
+            size_of::<XNonce>() + plaintext.len() + size_of::<Tag>()
+        );
+
+        // Valid decryption must succeed and yield correct plaintext
         let decrypted_text =
             decrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), ciphertext.clone()).unwrap();
-
-        // decrypted text must be equal to the original plaintext
         assert_eq!(decrypted_text, plaintext);
 
-        // must fail with a wrong domain
+        // Must fail on an incorrect domain
         assert!(decrypt_bytes_integral_nonce(&cipher, b"wrong_domain".to_vec(), ciphertext.clone()).is_err());
 
-        // must fail without nonce
-        assert!(decrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), ciphertext[0..12].to_vec()).is_err());
-
-        // must fail without mac
+        // Must fail with an evil nonce
+        let ciphertext_with_evil_nonce = ciphertext
+            .clone()
+            .splice(0..size_of::<XNonce>(), [0u8; size_of::<XNonce>()])
+            .collect();
+        assert!(decrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), ciphertext_with_evil_nonce).is_err());
+
+        // Must fail with malleated ciphertext
+        let ciphertext_with_evil_ciphertext = ciphertext
+            .clone()
+            .splice(
+                size_of::<XNonce>()..(ciphertext.len() - size_of::<Tag>()),
+                vec![0u8; plaintext.len()],
+            )
+            .collect();
+        assert!(
+            decrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), ciphertext_with_evil_ciphertext).is_err()
+        );
+
+        // Must fail with malleated authentication tag
+        let ciphertext_with_evil_tag = ciphertext
+            .clone()
+            .splice((ciphertext.len() - size_of::<Tag>())..ciphertext.len(), vec![
+                0u8;
+                size_of::<
+                    Tag,
+                >(
+                )
+            ])
+            .collect();
+        assert!(decrypt_bytes_integral_nonce(&cipher, b"correct_domain".to_vec(), ciphertext_with_evil_tag).is_err());
+
+        // Must fail if truncated too short
         assert!(decrypt_bytes_integral_nonce(
             &cipher,
-            b"correct_domain".to_vec(),
-            ciphertext[0..ciphertext.len().saturating_sub(32)].to_vec()
+            b"correct domain".to_vec(),
```

Comment:
I think we need the `_` here, as `b"correct_domain"` is used throughout the test
```suggestion
            b"correct_domain".to_vec(),
```

Reply:
Yep! This was a typo.